Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2024 Apple Inc. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_SWIFTUI && compiler(>=6.0)
+
+import Foundation
+internal import WebKit_Internal
+
+@_spi(Private)
+public struct URLScheme_v0: Sendable, Hashable, Equatable {
+    let rawValue: String
+
+    public init?(_ rawValue: String) {
+        guard let canonicalScheme = WKWebViewConfiguration._canonicalScheme(forURLScheme: rawValue) else {
+            return nil
+        }
+
+        self.rawValue = canonicalScheme
+    }
+}
+
+@_spi(Private)
+public enum URLSchemeTaskResult_v0: Sendable {
+    case response(URLResponse)
+
+    case data(Data)
+}
+
+@_spi(Private)
+public protocol URLSchemeHandler_v0 {
+    associatedtype TaskSequence: AsyncSequence<URLSchemeTaskResult_v0, any Error>
+
+    func reply(for request: URLRequest) -> TaskSequence
```

Comment:
`TaskSequence` does kinda run afoul of `Swift.Task`. We might want to workshop this a bit more.

I also wonder if we should call this something like
```swift
func handle(_ request: URLRequest) -> Result
```
to lean in to the "handler" part of the name, but probably fine as-is for now